ethtool_link_settings.go

```diff
@@ -0,0 +1,327 @@
+}
+
+// GetLinkSettings retrieves link settings, preferring ETHTOOL_GLINKSETTINGS and falling back to ETHTOOL_GSET.
+// Uses a single ioctl call with the maximum expected buffer size.
+func (e *Ethtool) GetLinkSettings(intf string) (*LinkSettings, error) {
+	// 1. Attempt ETHTOOL_GLINKSETTINGS with max buffer size
+	var req ethtoolLinkSettingsRequest
+	req.Settings.Cmd = ETHTOOL_GLINKSETTINGS
+	// Provide the maximum expected nwords based on our constant
+	req.Settings.LinkModeMasksNwords = int8(MAX_LINK_MODE_MASK_NWORDS)
+
+	err := e.ioctl(intf, uintptr(unsafe.Pointer(&req)))
+	attemptFallback := false
+	fallbackReason := ""
+
+	// Condition 1: ioctl returned EOPNOTSUPP
+	if errno, ok := err.(syscall.Errno); ok && errno == unix.EOPNOTSUPP {
+		attemptFallback = true
+		fallbackReason = "EOPNOTSUPP"
+	} else if err == nil {
+		// Condition 2: ioctl succeeded, but nwords might be invalid or buffer too small
+		nwords := int(req.Settings.LinkModeMasksNwords)
+		switch {
+		case nwords <= 0 || nwords > MAX_LINK_MODE_MASK_NWORDS:
+			// Sub-case 2a: Invalid nwords -> fallback
+			fmt.Printf("Warning: GLINKSETTINGS succeeded but returned invalid nwords (%d), attempting fallback to GSET\n", nwords)
+			attemptFallback = true
+			fallbackReason = "invalid nwords from GLINKSETTINGS"
+		case 3*nwords > len(req.Masks):
+			// Sub-case 2b: Buffer too small -> error
+			return nil, fmt.Errorf("kernel requires %d words for GLINKSETTINGS, buffer only has space for %d (max %d)", nwords, len(req.Masks)/3, MAX_LINK_MODE_MASK_NWORDS)
+		default:
+			// Sub-case 2c: Success (nwords valid and buffer sufficient)
+			results := &LinkSettings{
+				Speed:                req.Settings.Speed,
+				Duplex:               req.Settings.Duplex,
+				Port:                 req.Settings.Port,
+				PhyAddress:           req.Settings.PhyAddress,
+				Autoneg:              req.Settings.Autoneg,
+				MdixSupport:          req.Settings.MdixSupport,
+				EthTpMdix:            req.Settings.EthTpMdix,
+				EthTpMdixCtrl:        req.Settings.EthTpMdixCtrl,
+				Transceiver:          req.Settings.Transceiver,
+				MasterSlaveCfg:       req.Settings.MasterSlaveCfg,
+				MasterSlaveState:     req.Settings.MasterSlaveState,
+				SupportedLinkModes:   parseLinkModeMasks(req.Masks[0*nwords : 1*nwords]),
+				AdvertisingLinkModes: parseLinkModeMasks(req.Masks[1*nwords : 2*nwords]),
+				LpAdvertisingModes:   parseLinkModeMasks(req.Masks[2*nwords : 3*nwords]),
+				Source:               "GLINKSETTINGS",
+			}
+			if results.Speed == SPEED_UNKNOWN {
+				results.Speed = 0
+			}
+			return results, nil
+		}
+	} else if err != nil {
+		// Condition 3: ioctl failed with an error other than EOPNOTSUPP
+		// No fallback in this case.
+		return nil, fmt.Errorf("ETHTOOL_GLINKSETTINGS ioctl failed: %w", err)
+	}
+
+	// If we reach here, it means attemptFallback is true
+	if attemptFallback {
+		// Fallback to ETHTOOL_GSET using e.CmdGet
+		var cmd EthtoolCmd
```

Comment:
You could move this block within the if statement line 73. It will remove the need of the variable attemptFallback & fallbackReason